### DISCUSSION TITLE
Cypress Cloudinary file upload returns 400 - "Missing required parameter - file"

### DISCUSSION THREAD

**Discussion by pablohassan1** [cypress-io/cypress#15016] [Questions and Help]
I am testing image upload through Cloudinary iframe. Everything works as expected until the point when I submit the upload. The XHR request fails with 400 (Bad Request) status code providing error: "Missing required parameter - file". It works fine when testing manually, I compared both manual and Cypress requests and they seem identical.

Can someone suggest a solution? 
 
```
describe("Upload image to Cloudinary", () => {
  const iframeUrl = "url";

  before(function () {
    cy.fixture("users").then(function (users) {
      cy.keycloakLogin(users.user1.username, users.user1.password);
    });
  });

  after(() => {
    cy.keycloakLogout();
  });

  it("should be able to upload the image through Cloudinary iframe", function () {
    // navigate to appropiate app section
    navigateToProfile();

    //open Cloudinary iframe
    cy.findByText("Upload image").click({ force: true });

    // verify the iframe has loaded
    cy.frameLoaded("iframe[data-test='uw-iframe']", {
      url: iframeUrl,
    }).as("iframe");

    // upload the image
    cy.iframe("@iframe", {
      url: iframeUrl,
    }).then((iframe) => {
      cy.wrap(iframe).find(".cloudinary_fileupload").attachFile("selfie.jpeg");

      // it works well so far, the error occurs after the below step
      cy.wrap(iframe).find("button[type='submit']").click();
    });
  });
});
```

**Comment by portyu9** [comment]
Hi @pablohassan1 

The `400 "Missing required parameter - file"` response means Cloudinary received the request, but the multipart request did not contain a usable `file` part.

I would first replace the old `attachFile()` plugin with Cypress's native `selectFile()` and verify the browser input actually contains the file before submitting:

```js
cy.iframe('@iframe', { url: iframeUrl })
  .find('input[type="file"].cloudinary_fileupload')
  .selectFile('cypress/fixtures/selfie.jpeg', { force: true })
  .then(($input) => {
    expect($input[0].files).to.have.length(1)
    expect($input[0].files[0].name).to.eq('selfie.jpeg')
  })
```

Then wait for a **real Cloudinary UI state** showing that the file has been processed/selected before clicking Submit, rather than adding an arbitrary delay:

```js
cy.iframe('@iframe', { url: iframeUrl })
  .find('button[type="submit"]')
  .should('be.enabled')
  .click()
```

If `input.files` contains `selfie.jpeg` but Cloudinary's POST still has no `file` multipart field, the problem is between the simulated file-selection event and Cloudinary's uploader—not the Cloudinary API parameters. Inspect the request's **multipart Form Data**, not just its headers.

One other important point: if the Cloudinary iframe is cross-origin, direct iframe DOM automation is not fully supported by Cypress. In that case I would avoid relying on `cypress-iframe`/`chromeWebSecurity: false` as the long-term solution and either test your application's Cloudinary callback/integration separately or use browser-level automation for the third-party iframe.

So I would not change the Cloudinary API request manually—the first thing to prove is that Cypress actually selected a native file and that the widget consumed it before Submit.

**Reply by portyu9** [reply]
@pablohassan1 Please mark this answer as accepted if it helped, thank you.
